src/rgw/rgw_op.cc

```diff
@@ -3726,13 +3727,49 @@ static CompressorRef get_compressor_plugin(const req_state *s,
   return Compressor::create(s->cct, alg);
 }
 
+class AsyncMD5 {
+  MD5* hash;
+  std::future<void> last;
+  Throttle ongoing_ops;
+public:
+  AsyncMD5(CephContext *cct) : hash(new MD5()), ongoing_ops(cct,
+    "ongoing_md5_ops", cct->_conf->rgw_async_md5_ops) {}
+  ~AsyncMD5() {
+    delete hash;
+    hash = nullptr;
+  }
+
+  // NOTE: data buffer will be freed within this function
+  void Update(char *data, size_t len) {
+    ongoing_ops.get(1);
+
+    last = std::async([&](std::future<void>&& last, char* data,
```

Comment:
this pattern with `std::async()` and `std::future` looks simple and elegant, but i do have some concerns here

first is that rgw requests can run asynchronously as coroutines in a `boost::asio::io_context`, and we want to avoid blocking on a condition variable (either in `Throttle::get()` or `std::future::get()`) - instead, we should suspend the coroutine so this thread can resume work on something else until the result is ready. you can find some examples of this in `RGWReshardWait::wait(optional_yield y)` and `RGWHTTPClient::wait(optional_yield y)`

[this overload](https://en.cppreference.com/w/cpp/thread/async) of `std::async()` doesn't take a launch policy, and
> Behaves as if called with `policy` being `std::launch::async | std::launch::deferred`. In other words, f might be executed in another thread or it might be run synchronously when the resulting `std::future` is queried for a value.

as far as i know, `std::launch::async` doesn't give us any guarantees about the number or lifetime of its background threads, or the order of execution for these tasks. if it _does_ limit the number of threads _and_ allows out-of-order execution, this pattern could lead to deadlocks because each task blocks on the result of the previous task with `std::future::get()`

Reply:
> you can find some examples of this in RGWReshardWait::wait(optional_yield y) and RGWHTTPClient::wait(optional_yield y)

ok, thank you.

> if it does limit the number of threads and allows out-of-order execution, this pattern could lead to deadlocks because each task blocks on the result of the previous task with std::future::get()

@cbodley 
 I don't know under which condition this will lead to deadlocks. Can you give an example case?
![image](https://user-images.githubusercontent.com/20939112/127838258-6e2258d1-1b20-4bda-aff7-edb322451591.png)

Reply:
> > if it does limit the number of threads and allows out-of-order execution, this pattern could lead to deadlocks because each task blocks on the result of the previous task with std::future::get()
> 
> I don't know under which condition this will lead to deadlocks. Can you give an example?

taking this example to an extreme, consider an implementation that uses a thread pool with a single thread. we upload an object, and `AsyncMD5` creates a sequence of tasks `A->B->C->D`. if this thread pool allows out-of-order execution, then it may execute `B` before `A`. task `B` will block waiting for the result of `A`, but task `A` can never run because there's only the single thread

as the number of threads incease, deadlock becomes far less likely. but i think we're better off handling the threading manually to guarantee that it won't. for example, with a scheduler that's aware of these dependencies, and doesn't schedule a task until it's ready to run. we should also be able to combine the two separate locks (`Throttle` and `std::future`) into one

ultimately i think we're either going to need SIMD or large batches to see real wins here, to make up for the added overhead of thread synchronization

Reply:
> ultimately i think we're either going to need SIMD or large batches to see real wins here, to make up for the added overhead of thread synchronization

I'm with you on that. Thank you for your suggestions.